File: test-suite-caffeine-kotlin/src/test/kotlin/io/micronaut/cache/RemovalListenerImpl.kt
package io.micronaut.cache

import com.github.benmanes.caffeine.cache.RemovalCause
import com.github.benmanes.caffeine.cache.RemovalListener
import jakarta.inject.Singleton

// tag::clazz[]
@Singleton
class RemovalListenerImpl internal constructor(private val handler: MyRemovalHandler) : RemovalListener<String?, Int?> {

    override fun onRemoval(key: String?, value: Int?, cause: RemovalCause) {
        handler.handle(key!!, value!!, cause)
    }
}
// end::clazz[]

File: test-suite-caffeine-python/src/test/python/micronaut/cache/docs/RemovalListenerImpl.py
from com.github.benmanes.caffeine.cache import RemovalCause, RemovalListener
from jakarta.inject import Singleton

from .MyRemovalHandler import MyRemovalHandler


# tag::clazz[]
@Singleton
class RemovalListenerImpl(RemovalListener[str, int]):

    def __init__(self, handler: MyRemovalHandler):
        self.handler = handler

    def onRemoval(self, key: str | None, value: int | None, cause: RemovalCause) -> None:
        self.handler.handle(key, value, cause)
# end::clazz[]
